torch/distributed/rpc/backend_registry.py

```diff
@@ -243,6 +243,80 @@ def _tensorpipe_exchange_and_check_all_device_maps(
 
     return reverse_device_maps, my_devices
 
+def _tensorpipe_check_local_device_maps(name, options):
+    # Check local devices in device_maps and devices are all valid.
+    local_devices = set(options.devices) if options.devices else set()
+    device_maps = options.device_maps
+    for worker_name in device_maps:
+        device_map = device_maps[worker_name]
+        key_set = set(device_map.keys())
+        val_set = set(device_map.values())
+        if not all([
+            len(key_set) == len(device_map),
+            len(val_set) == len(device_map),
+        ]):
+            raise ValueError(
+                f"Invalid device_map configuration for {worker_name}, "
+                f"not 1-to-1 mapping:\ndevice_maps = {device_map}"
+            )
+        local_devices.update(key_set)
+
+    if not all(
+        (0 <= d.index < torch.cuda.device_count() if d.type == "cuda" else True)
+        for d in local_devices
+    ):
+        raise ValueError(
+            f"Invalid device in TensorPipe options on {name}:\n"
+            f"device_maps = {options.device_maps},\n"
+            f"devices = {options.devices}"
+        )
+
+# detect if any worker has invalid device_map configurations, and return
+# names of failed workers
+def _tensorpipe_check_remote_device_maps(agent, options):
+    device_maps = options.device_maps
+    if device_maps is None:
+        device_maps = {}
+
+    def check_one_worker(name, device_maps, all_device_counts):
+        device_count = all_device_counts[name]
+        wrong_worker_names = set(device_maps) - set(all_device_counts)
+        if wrong_worker_names:
+            raise ValueError(f"Wrong worker names: {wrong_worker_names}")
+        for remote_name in all_device_counts:
+            remote_device_count = all_device_counts[remote_name]
+            if remote_name in device_maps:
+                device_map = device_maps[remote_name]
+                val_set = set(device_map.values())
+                if not all(
+                    (0 <= d.index < remote_device_count if d.type == "cuda" else True)
+                    for d in val_set
+                ):
+                    raise ValueError(
+                        f"Invalid device_map configuration on {name} "
+                        f"for {remote_name}, remote device out of range:\n"
+                        f"device_maps = {device_maps}"
+                    )
+
+    gathered = api._all_gather([torch.cuda.device_count(), device_maps])
```

Comment:
hmm, how does this work? It uses `_ALL_WORKER_NAMES` to do allgather?

Reply:
Removed this function in the next commit